Implement a contao_ variant of the html_attr escaper

m-vo opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | see #3200 
| Docs PR or issue | -

This introduces a new `contao_html_attr` strategy that replaces `html_attr`. 

Even if we do not want this change we should port the fix in `ContaoEscaperNodeVisitor#isEscaperFilterExpressionWithHtmlStrategy` (support `e` or `escape` as name).

/cc @ausi

m-vo committed:
support different contao escapers + implement html_attr variant

m-vo committed:
explicitly mock adapter

m-vo commented on core-bundle/tests/Twig/Interop/ContaoEscaperNodeVisitorTest.php:
```diff
@@ -97,16 +99,47 @@ public function testUppercaseEntities(): void
+        $controller = $this->getMockBuilder(Adapter::class)
+            ->disableOriginalConstructor()
+            ->addMethods(['replaceInsertTags'])
+            ->setMockClassName(Controller::class)
```
I have no idea why this is necessary. A normal mock works (but has the wrong type in `getAdapter`) and setting the mock class name works. If I don't do either, the adapter is `null` when multiple tests are run.

IMHO looks a bit like a PHPUnit caching issue.

leofeyer replied:
This triggers the following warnings on my system:

```
There were 5 warnings:

1) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testEscapesEntities
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

2) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testDoesNotDoubleEncode
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

3) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testHandlesFiltersAndFunctions
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

4) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testUppercaseEntities
Class "Contao\Controller" already exists.

/Users/leofeyer/Sites/contao/managed-edition/vendor/contao/contao/vendor/phpunit/phpunit/phpunit:76

5) Contao\CoreBundle\Tests\Twig\Interop\ContaoEscaperNodeVisitorTest::testHtmlAttrFilter
Class "Contao\Controller" already exists.
```

m-vo replied:
Ah, then this might be an autoloading issue. :thinking:

m-vo replied:
This was a tough one. Fixed in 4dd61be7b.

Generating and executing Twig sources did not play well with PHPUnits reference handling. I rewrote the code to use old school `Controller::replaceInsertTags()` instead and the tests to mock and set a container with `System::getContainer` and real insert tag listeners.

ausi commented on core-bundle/src/Twig/Interop/ContaoEscaper.php:
```diff
@@ -43,10 +60,85 @@ public function __invoke(Environment $environment, $string, ?string $charset): s
+        $string = (string) $string;
+
+        // Replace insert tags before '{' and '}' get encoded
+        $string = $this->framework->getAdapter(Controller::class)->replaceInsertTags($string);
```
```suggestion
        $string = $this->framework->getAdapter(Controller::class)->replaceInsertTags($string, false);
```

`$blnCache` needs to be `false` I think, otherwise uncachable insert tags will not get replaced.

m-vo committed:
disable insert tag caching

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

leofeyer committed:
CS

m-vo committed:
do not mock framework, stay old school

this does not work reliably as Twig sources are generated and executed by the tests - the references to the mocked injected services are then lost under some circumstances

leofeyer committed:
Fix the tests

leofeyer commented:
Thank you @m-vo.
